documentation/design-docs/ipc-protocol.md

```diff
@@ -0,0 +1,555 @@
+
+### Commands
+
+Commands are a `command_set` and a `command_id`.  A `command_set` is analogous to a namespace for `command_id`s.  The typical grouping is by service running on the Diagnostic Server, e.g., there is a `command_set` for EventPipe.  This allows multiple services to have the same `command_id`s without clashing.  The combination of a `command_set` and a `command_id` encodes the Command being invoked on the Diagnostic Server.
+
+The current set of `command_set`s and `command_id`s are listed below:
+
+```c
+enum class CommandSet : uint8_t
+{
+    // reserved = 0x00,
+    Miscellaneous = 0x01,
+    EventPipe     = 0x02,
+    // future
+
+    Server = 0xFF,
+};
+```
+
+```c
+enum class ServerCommandId : uint8_t
+{
+    OK    = 0x00,
+    Error = 0xFF,
+};
+```
+
+Commands may use the generic `{ magic="DOTNET_IPC_V1"; size=20; command_set=0xFF (Server); command_id=0x00 (OK); reserved = 0x0000; }` to indicate success rather than having a command specific success `command_id`.  Similarly, Commands may use the `command_set=0xFF (Server); command_id=0xFF (Error);` to generically indicate an error has occurred.
+
+For example, the Command to start a stream session with EventPipe would be `0x0202` made up of `0x02` (the `command_set` for EventPipe) and `0x02` (the `command_id` for CollectTracing).
+
+#### EventPipe Commands
+
+```c
+enum class EventPipeCommandId : uint8_t
+{
+    // reserved = 0x00,
+    StopTracing    = 0x01, // stop a given session
+    CollectTracing = 0x02, // create/start a given session
+}
+```
+EventPipe Payloads are encoded with the following rules:
+```c
+// X, Y, Z means encode bytes for X followed by bytes for Y followed by bytes for Z
+// uint = 4 little endian bytes
+// ulong = 8 little endian bytes
+// wchar = 2 little endian bytes, UTF16 encoding
+// array<T> = uint length, length # of Ts
+// string = (array<wchar> where the last wchar must = 0) or (length = 0)
+```
+
+The `CollectTracing` Command consumes the following Payload:
+```
+Payload
+{
+    ulong PID,
+    uint circularBufferMB,
+    string outputPath,
+    array<provider_config> providers
+}
+
+provider_config 
+{
+    ulong keywords,
+    uint logLevel,
+    string provider_name,
+    string filter_data (optional)
+}
+```
+
+and responds with a Payload of
+```c
+Payload
+{
+    ulong sessionId
+}
+```
+Followed by an Optional Continuation of a `netperf` format stream of events.
+
+The `StopTracing` Command consumes the following Payload:
+```c
+Payload
+{
+   ulong sessionId
+}
+```
+
+and responds with a Payload of
+```c
+Payload
+{
+   ulong sessionId
+}
+```
+
+### Errors
+
+Errors are mostly Command specific, but there are some generic Diagnostic Server errors described in the code snippet below.
+
+```c++
+enum class DiagnosticServerErrorCode : uint32_t
+{
+    OK                = 0x00000000,
+    BadEncoding       = 0x00000001,
+    UnknownCommandSet = 0x00000002,
+    UnknownCommandId  = 0x00000003,
+    UnknownVersion    = 0x00000004,
```

Comment:
Nit: Elsewhere you referred to it as 'Magic' so I'd recommend keeping that terminology.